### PR TITLE
Added missing init data for the sale package, refactored, and added tests for the invoice package

### PR DESCRIPTION
**Pull request by alexandraYesbabylon**

### CODE REVIEW

File: finance/actions/invoice/do-emit.php [hunk 1]
```diff
@@ -26,19 +26,14 @@
...
 
-/**
- * @var \equal\php\Context                  $context
- * @var \equal\orm\ObjectManager            $orm
- * @var \equal\auth\AuthenticationManager   $auth
- */
-list($context, $orm, $auth) = [$providers['context'], $providers['orm'], $providers['auth']];
 
-// emit the invoice : changing status will trigger an invoice number assignation
+$context = [$providers['context']];
```
Comment (lucaslaurent04) [inline finance/actions/invoice/do-emit.php:33]: Remove brackets :)   @alexandraYesbabylon

File: finance/classes/accounting/AccountingRuleLine.class.php [hunk 2]
```diff
@@ -57,14 +57,11 @@ public static function getColumns() {
...
         $result = [];
-
-        $res = $om->read(get_called_class(), $oids, ['account_id.code']);
-        if($res > 0 && count($res)) {
-            foreach($res as $oid => $odata) {
-                $result[$oid] = $odata['account_id.code'];
-            }
+        $self->read(['account_id' => ['code']]);
+        foreach($self as $id => $line) {
+            $result[$id] = $line['code'];
```
Comment (lucaslaurent04) [inline finance/classes/accounting/AccountingRuleLine.class.php:64]:
```php
$result[$id] = $line['account_id']['code'];
```

File: finance/classes/accounting/Invoice.class.php [hunk 3]
```diff
@@ -197,40 +197,34 @@ public static function getColumns() {
...
+        foreach($self as $id => $invoice) {
+            $result[$id] = false;
+            if($invoice['status'] != 'invoice') {
+                continue;
+            }
+            if($invoice['price'] == 0) {
+                $result[$id] = true;
+                continue;
+            }
+            if($invoice['funding_id'].['paid_amount'] && $invoice['funding_id'].['paid_amount'] == $invoice['price']) {
```
Comment (lucaslaurent04) [inline finance/classes/accounting/Invoice.class.php:212]: A dot left

$invoice['funding_id'].['paid_amount']      should be    -> $invoice['funding_id']['paid_amount']

File: finance/classes/accounting/Invoice.class.php [hunk 4]
```diff
@@ -197,40 +197,34 @@ public static function getColumns() {
...
-    public static function calcPaymentReference($om, $oids, $lang) {
+    public static function calcPaymentReference($self): array {
         $result = [];
-        $invoices = $om->read(get_called_class(), $oids, ['number']);
-        foreach($invoices as $oid => $invoice) {
-            $number = intval($invoice['number']);
-            // arbitrary value : 155 for balance (final) invoice
-            $code_ref = 155;
-            $result[$oid] = self::_get_payment_reference($code_ref, $number);
+        $self->read(['invoice_number']);
```
[the remark this answers is not in the document]
Comment (alexandraYesbabylon) [reply finance/classes/accounting/Invoice.class.php]: [Fixed typing bugs in finance packages](https://github.com/discope-pms/discope/pull/5/commits/8b282d37f7f40de461b09c5c0efff46dbee46e99)

**Pull request merged**